hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/joiner/DefaultEntityJoinerBuilder.java

```diff
@@ -241,26 +246,58 @@ private String getEntityType(SelectedField entityField) {
       return buildIdFilter(context, entityType, entityIdsToFilter)
           .flatMap(
               filterArguments ->
-                  resultSetRequestBuilder.build(
-                      context,
-                      entityType,
-                      entityIdsToFilter.size(),
-                      ZERO_OFFSET,
-                      new InstantTimeRange(),
-                      List
-                          .<AttributeAssociation<AggregatableOrderArgument>>
-                              of(), // Order does not matter
-                      filterArguments,
-                      this.entityFieldsByType.get(entityType).stream(),
-                      Optional.empty()))
-          .map(resultSetRequest -> new DefaultEntityRequest(entityType, resultSetRequest));
+                  buildEntityRequest(
+                      context, entityType, entityIdsToFilter.size(), filterArguments));
+    }
+
+    private Single<EntityRequest> buildEntityRequest(
+        GraphQlRequestContext context,
+        String entityType,
+        int entityIdsToFilterSize,
+        List<AttributeAssociation<FilterArgument>> filterArguments) {
+      return zip(
+          resultSetRequestBuilder.build(
+              context,
+              entityType,
+              entityIdsToFilterSize,
+              ZERO_OFFSET,
+              new InstantTimeRange(),
+              List.<AttributeAssociation<AggregatableOrderArgument>>of(), // Order does not matter
+              filterArguments,
+              this.entityFieldsByType.get(entityType).stream(),
+              Optional.empty()),
+          attributeRequestBuilder.buildForKey(context, entityType, LABELS_KEY_NAME),
+          (resultSetRequest, labelsAttributeRequest) ->
+              new DefaultEntityRequest(
+                  entityType,
+                  resultSetRequest,
+                  canFetchLabels(entityType)
+                      ? Optional.of(new DefaultLabelRequest(labelsAttributeRequest))
+                      : Optional.empty()));
     }
 
     private Single<List<AttributeAssociation<FilterArgument>>> buildIdFilter(
         GraphQlRequestContext context, String entityScope, Collection<String> entityIds) {
       return filterRequestBuilder.build(
           context, entityScope, Set.of(new EntityIdFilter(entityIds, entityScope)));
     }
+
+    private boolean canFetchLabels(String entityType) {
```

Comment:
This is mostly duplicated, which isn't great considering we want to spread this to more places, too.
I wanted to suggest moving label request building into its own builder so I started writing a snippet... and I got carried away. I pushed it to a separate branch so you can either reference it to get an idea of what I was trying to suggest, or just merge it into this branch as is - although I didn't update tests.

https://github.com/hypertrace/hypertrace-graphql/pull/103

Reply:
Your 15 minutes saved an hour of my effort :-)

Merged your changes to this branch, and manually verified that changes work fine.